File: edc/org.eclipse.cdt.debug.edc.tcf.extension/src/org/eclipse/cdt/debug/edc/tcf/extension/AgentException.java
/*******************************************************************************
 * Copyright (c) 2009 Nokia and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * Nokia - Initial API and implementation
 *******************************************************************************/

package org.eclipse.cdt.debug.edc.tcf.extension;

/**
 * Checked exception used only within a TCF agent. One main purpose is to
 * support cascading of error messages along a call chain.
 * 
 * @author LWang
 * 
 */
public class AgentException extends Exception {

	private static final long serialVersionUID = 8073163639364456392L;

	private final String fMessage;

	/**
	 * @param fMessage
	 */
	public AgentException(String message) {
		super(message);
		fMessage = null;
	}

	/**
	 * @param cause
	 */
	public AgentException(Throwable cause) {
		super(cause);
		fMessage = null;
	}

	/**
	 * @param fMessage
	 * @param cause
	 */
	public AgentException(String message, Throwable cause) {
		super(message, cause);
		fMessage = (cause != null) ? message + "\n=>Underlying cause: " + cause.getMessage() : null;
	}

	@Override
	public String getMessage() {
		return (fMessage != null) ? fMessage : super.getMessage();
	}
}

File: edc/org.eclipse.cdt.debug.edc.windows.agent/src/common_agent/AgentException.cpp
/*******************************************************************************
 * Copyright (c) 2009 Nokia and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 * Nokia - Initial API and implementation
 *******************************************************************************/
#include "AgentException.h"

AgentException::AgentException() {
}

AgentException::AgentException(const std::string& msg) {
	message = msg;
	cause = NULL;
}

AgentException::AgentException(const char* msg) {
	message = msg;
	cause = NULL;
}

AgentException::AgentException(const char* msg, AgentException* cause) {
	message = msg;
	this->cause = cause;
}

AgentException::~AgentException() throw () {
	if (cause != NULL)
		delete cause;
}

const char* AgentException::what() const throw () {
	if (cause != NULL) {
		std::string m = message;
		m += "\n\t";
		m += cause->what();
		return m.c_str();
	}

	return message.c_str();
}
